run_flax_speech_recognition_seq2seq_streaming_review.py

```diff
@@ -0,0 +1,1391 @@
+            "num_train_dataset_workers": f"{num_workers}",
+            "num_hosts": f"{num_of_hosts}",
+            "total_num_training_examples": f"{data_args.num_train_steps * train_batch_size:,}",
+            "steps_per_epoch": "_To be computed after first epoch_",
+            "num_beams": model_args.num_beams,
+        },
+        # TODO: Adapt https://github.com/huggingface/transformers/blob/main/src/transformers/modelcard.py#L855
+        # "hyperparameters": training_args.to_sanitized_dict()
+    }   
+        
+    if training_args.gradient_accumulation_steps > 1:
+        training_summary["hyperparameters"]["gradient_accumulation_steps"] = f"{training_args.gradient_accumulation_steps:,}"
+        training_summary["hyperparameters"]["effective_total_train_batch_size"] = f"{train_batch_size * training_args.gradient_accumulation_steps:,}"
+    
+    if model_args.dropout or model_args.attention_dropout or model_args.activation_dropout or model_args.encoder_dropout or model_args.decoder_dropout:
+        training_summary["hyperparameters"]["dropout"] = True
+        if model_args.dropout:
+            training_summary["hyperparameters"]["dropout_probability"] = model_args.dropout
+        if model_args.attention_dropout:
+            training_summary["hyperparameters"]["attention_dropout_probability"] = model_args.attention_dropout
+        if model_args.activation_dropout:
+            training_summary["hyperparameters"]["activation_dropout_probability"] = model_args.activation_dropout
+        if model_args.encoder_dropout:
+            training_summary["hyperparameters"]["encoder_dropout_probability"] = model_args.encoder_dropout
+        if model_args.decoder_dropout:
+            training_summary["hyperparameters"]["decoder_dropout_probability"] = model_args.decoder_dropout
+
+    # Create README if it does not exist
+    readme = output_dir / "README.md"
+    if not readme.exists():
+        readme.write_text(TrainingSummary(**training_summary).to_model_card())
+    
+    # ======================== Training ================================
+    train_start = time.time()
+
+    train_metrics = []
+    epoch = 0
+    if training_args.do_train:
+        train_dataset = vectorized_datasets["train"].shuffle(seed=training_args.seed, buffer_size=data_args.shuffle_buffer_size)
+        # Split by node
+        train_dataset = split_dataset_by_node(train_dataset, rank=current_host_idx, world_size=num_of_hosts)   
+    
+        if train_dataset.n_shards < data_args.preprocessing_num_workers:
+            num_workers = train_dataset.n_shards
+
+        logger.info(f"  Number of train dataset workers = {num_workers} {'(Capped by the number of dataset shards)' if train_dataset.n_shards < data_args.preprocessing_num_workers else ''} {'(ADVICE: In most cases you will speed up training considerably if you increase the value of --preprocessing_num_workers!)' if num_workers < 10 else ''}")
```

Comment:
Probably we want to reserve `data_args.preprocessing_num_workers` for dataset.map, and `training_args.dataloader_num_workers` for the data loader https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments.dataloader_num_workers

It's ok if we don't have the `logger.info` warning here - it's the same with `preprocessing_num_workers` where we'd want to advise the user outside of the script to bump up this value to the max possible one

Reply:
We removed the warning in the `logger.info()` and changed from `data_args.preprocessing_num_workers` to `training_args.dataloader_num_workers`. We still believe that the default for `training_args.dataloader_num_workers` should not be 0 when training.

Reply:
I definitely see your point, but having `dataloader_num_workers > 0` relies on the data being properly formatted for streaming mode. Even for the most popular ASR dataset ([librispeech_asr](https://huggingface.co/datasets/librispeech_asr)), this isn't the case, since the original data is not sharded into chunks (but rather saved as one massive file).

In short, we can't guarantee that every dataset will support `dataloader_num_workers > 0`. However, what's quite nice about HF Datasets is that it will automatically reduce the number of workers if there aren't enough shards, so probably actually having `dataloader_num_workers > 0` is ok since we'll always drop down to the largest possible num workers